components/raftstore/src/store/worker/read.rs

```diff
@@ -165,13 +173,126 @@ pub struct ReadDelegate {
     pub track_ver: TrackVer,
 }
 
+/// CachedReadDelegate is a wrapper the ReadDelegate and kv_engine. LocalReader
+/// dispatch local read requests to ReadDeleage according to the region_id where
+/// ReadDelegate needs kv_engine to read data or fetch snapshot.
+pub struct CachedReadDelegate<E>
+where
+    E: KvEngine,
+{
+    delegate: Arc<ReadDelegate>,
+    kv_engine: E,
+}
+
+impl<E> Deref for CachedReadDelegate<E>
+where
+    E: KvEngine,
+{
+    type Target = ReadDelegate;
+
+    fn deref(&self) -> &Self::Target {
+        self.delegate.as_ref()
+    }
+}
+
+impl<E> Clone for CachedReadDelegate<E>
+where
+    E: KvEngine,
+{
+    fn clone(&self) -> Self {
+        CachedReadDelegate {
+            delegate: Arc::clone(&self.delegate),
+            kv_engine: self.kv_engine.clone(),
+        }
+    }
+}
+
+/// ReadDelegateExt combines some LocalReader's fields for temporary usage.
+pub struct ReadDelegateExt<'a, E>
+where
+    E: KvEngine,
+{
+    metrics: &'a mut ReadMetrics,
+    read_id: &'a mut ThreadReadId,
+    snap_cache: &'a mut Box<Option<Arc<E::Snapshot>>>,
+}
+
 impl Drop for ReadDelegate {
     fn drop(&mut self) {
         // call `inc` to notify the source `ReadDelegate` is dropped
         self.track_ver.inc();
     }
 }
 
+pub trait DelegateStore<E>: Send + Clone + 'static
```

Comment:
I feel like there are too many XxDelegate types around. `ReadExecutorProvider` is probably more meaningful here.